CIP-meta-assets (ERC20-like assets)/README.md

```diff
@@ -0,0 +1,428 @@
+
+```ts
+const AccounManagerRdmr = pstruct({
+    Mint: { // or Burn if `amount` is negative
+        amount: int
+    },
+    Transfer: {
+        to: PCredential.type,
+        amount: int
+    },
+    Receive: {},
+    ForwardCompatibility: {}
+});
+```
+
+The validation logic is different for each redeemer.
+
+#### Common operations and values
+
+Before proceeding with the redeemers validation logic here are some common operations and values between some of the redeemers.
+
+##### `ownUtxoRef`
+
+the `accountManager` contract is meant to be used only as spending validator.
+
+As succ we can extract the utxo being spent from the `ScriptPurpose` when constructed with the `Spending` contstructor and fail for the rest.
+
+```ts
+const ownUtxoRef = plet(
+    pmatch( ctx.purpose )
+    .onSpending(({ utxoRef }) => utxoRef)
+    ._( _ => perror( PTxOutRef.type ) )
+);
+```
+
+##### `validatingInput`
+
+is the input with `utxoRef` field equivalent to `ownUtxoRef`
+
+##### `ownCreds`
+
+from the `validatingInput`:
+
+```ts
+validatingInput.resolved.address.credential
+```
+
+##### `ownValue`
+
+from the `validatingInput`:
+
+```ts
+validatingInput.resolved.value
+```
+
+
+##### `isOwnOutput`
+
+given an output; we recongize the output as "own" if the attached credantials are equivalent to `ownCreds` 
+and the attached value includes an entry for the `currencySym` field in the specified in the datum.
+
+```ts
+const isOwnOutput = plet(
+    pfn([ PTxOut.type ], bool )
+    ( out => 
+        out.address.credential.eq( ownCreds )
+        // a single account manager contract might handle multiple tokens
+        .and(
+            out.value.some( ({ fst: policy }) => policy.eq( account.currencySym ) )
+        ) 
+    )
+);
+```
+
+##### `isOwnInput` 
+
+an input is "own" if the resolved field is "own";
+
+```ts
+const isOwnInput = plet(
+    pfn([ PTxInInfo.type ], bool )
+    ( input => isOwnOutput.$( input.resolved ) )
+);
+```
+
+##### `isOwnCurrencySym`
+
+given an asset policy we might want to know if it is the one being validated.
+
+this is done by comparing it with the one specified in the datum field
+
+```ts
+const isOwnCurrSym = plet( account.currencySym.eqTerm );
+```
+
+##### `outIncludesNFT`
+
+given a transaction output is useful to check if the value includes the NFT generated from the `accountFactory`;
+
+this is done by checking that at least one of the attached value's entry satisfies `isOwnCurrencySym` for the policy.
+
+```ts
+ const outIncludesNFT = plet(
+    pfn([ PTxOut.type ], bool )
+    ( out => out.value.some( entry => isOwnCurrSym.$( entry.fst ) ) )
+);
+```
+
+##### `ownOuts`
+
+transaction outputs filtered by `isOwnOutput`;
+
+##### `ownInputs`
+
+transaction inputs filtered by `isOwnInput`;
+
+##### updating the `Account` datum
+
+Between all the standard redeemers it must be checked that the datum fields `credentials`, `currencySym` and `state` remain unchanged compared to the current datum.
```

Comment:
Not intended to do a semantic review (just fixing typos), but I could not help to observe:

If there is a requirement of X to be unchanged, I would strongly suggest to mention it explicitly everywhere needed.

We had this same "ensure that all non mentioned parts of a datum remain unchanged" in the [Hydra specification](https://hydra.family/head-protocol/core-concepts/specification/) and it resulted in at least one security vulnerability where forgot that check.

Reply:
That makes a lot of sense, will make sure to report it where needed

Reply:
went through the specification and this seems to be reported where needed, let me know if you spot any inconsistencies!